app/assets/javascripts/components/course/course_approval.jsx

```diff
@@ -0,0 +1,354 @@
+import { difference, uniq } from 'lodash-es';
+import moment from 'moment';
+
+import PropTypes from 'prop-types';
+import Select from 'react-select';
+import CreatableSelect from 'react-select/creatable';
+import selectStyles from '../../styles/single_select';
+
+import { fetchSpecialUsers } from '../../actions/settings_actions';
+import { fetchAllCampaigns, addCampaign } from '../../actions/campaign_actions';
+import { removeTag, fetchAllTags, addTag } from '../../actions/tag_actions';
+import { addUser } from '../../actions/user_actions';
+import { getCourseApprovalStaff } from '../../selectors';
+import { STAFF_ROLE } from '../../constants';
+
+
+const CourseApproval = (props) => {
+  const [selectedWikiExpert, setSelectedWikiExpert] = useState({});
+  const [selectedCampaigns, setSelectedCampaigns] = useState([]);
+  const [selectedTags, setSelectedTags] = useState([]);
+  const [createdTagOption, setCreatedTagOption] = useState([]);
+  const [submitting, setSubmitting] = useState(false);
+
+  useEffect(() => {
+    props.fetchSpecialUsers();
+    props.fetchAllCampaigns();
+    props.fetchAllTags();
+  }, []);
+
+  useEffect(() => {
+    if (props.wikiEdStaff.length > 0) {
+      const wikiExpert = props.wikiEdStaff.find(user => user.role === 'wikipedia_expert');
+      const currentWikiExpert = props.wikiEdStaff.find(user => user.role === 'wikipedia_expert' && user.already_selected);
+
+      // Check if there exists a wiki expert who is already allotted a staff role. If yes, set that
+      // user as selectedWikiExpert. Else, set the first wikipedia expert user as the selectedWikiExpert.
+      setSelectedWikiExpert(
+        (currentWikiExpert !== null && currentWikiExpert !== undefined)
+          ? { value: currentWikiExpert.username, label: `${currentWikiExpert.username} (${currentWikiExpert.realname})` }
+          : { value: wikiExpert.username, label: `${wikiExpert.username} (${wikiExpert.realname})` }
+      );
+    }
+
+    if (props.allCampaigns.length > 0) {
+      setDefaultCampaign();
+    }
+
+    if (props.tags.length > 0) {
+      setSelectedTags(props.tags.map((tag) => {
+        return { value: tag.tag, label: tag.tag };
+      }));
+    }
+  }, [props.wikiEdStaff, props.allCampaigns, props.tags]);
+
+  const setProgramManager = () => {
+    const programManager = props.wikiEdStaff.find(user => user.role === 'classroom_program_manager');
+    return {
+      value: programManager.username,
+      label: `${programManager.username} (${programManager.realname})`
+    };
+  };
+
+  const setWikiExpertOptions = () => {
+    const wikiExperts = props.wikiEdStaff.filter(user => user.role === 'wikipedia_expert');
+    const options = wikiExperts.map((user) => {
+      return {
+        value: user.username,
+        label: `${user.username} (${user.realname})`
+      };
+    });
+    return options;
+  };
+
+  const setDefaultCampaign = () => {
+    // Try to infer the best suitable campaign from the start date of course.
```

Comment:
All this logic should be extracted to a separate file as a pure function, so that we can write simple tests for how a start date gets converted into a campaign slug.

Reply:
@ragesoss, is it okay to place this function inside `utils` folder in `app/assets/javascripts/components/course/` ?

Reply:
That sounds like the right place for it.